OpenQA/Isotovideo/Utils.pm

```diff
@@ -78,16 +81,61 @@ sub _fetch_new_refs ($clone_url, $cache_dir, $branch_args, $handle_output) {
     $handle_output->($?, qx{git -C "$cache_dir" branch --force $branch_args FETCH_HEAD 2>&1 2>&1}) if $branch_args;
 }
 
+sub _open_cache_index ($root_cache_dir, $index_file) {
+    my $index = -e $index_file ? decode_json($index_file->slurp) : {};
+    die "root is not an object\n" unless ref $index eq 'HASH';
+    for my $repo_path (keys %$index) {
+        my $repo = $index->{$repo_path};
+        die "entry '$repo' is invalid\n" unless looks_like_number($repo->{size}) && looks_like_number($repo->{last_use});
+        delete $index->{$repo} unless -e path($root_cache_dir, $repo_path);
+    }
+    return $index;
+}
+
+sub _determine_size ($dir, $handle_output) {
+    $handle_output->($?, my $du = qx{du -s "$dir"});
+    die "Unable to determine size of Git directory under '$dir': du returned '$du'\n" unless $du =~ /(\d+).*/;
+    return int($1);
+}
+
+sub limit_git_cache_dir ($root_cache_dir, $current_cache_dir, $current_relative_cache_dir, $handle_output) {
+    my $cache_dir_size = _determine_size($current_cache_dir, $handle_output);
+    my $index_file = path($root_cache_dir, 'index.json');
+    my $index_lock = _lock_cache_directory($index_file);
+    my $index = eval { _open_cache_index($root_cache_dir, $index_file) };
+    die "Unable to open index for Git caching under '$index_file': $@" if $@;
+    $index->{$current_relative_cache_dir} = {size => $cache_dir_size, last_use => time};
+    my $index_guard = scope_guard sub { $index_file->spew(encode_json($index)) };
+    return undef unless looks_like_number(my $limit = $bmwqemu::vars{GIT_CACHE_DIR_LIMIT});
```

Comment:
Then we'd skip the inexpensive part. I'd actually check `df` already before invoking `dh`. I'd still opening the JSON index to put an entry there but the size would be `undef`. I'd also skip rewriting the JSON file if such an entry already exists. (With this the code of course needs to handle the size being `undef` but that's not a big deal.)

Reply:
I don't understand "I'd actually check df already before invoking dh". With "dh" do you mean "du"? And are you saying that you are already checking free space somewhere where I didn't see it?

Reply:
Yes, I meant "du" and it is called in the function `_determine_size` I added in this PR. It is used for determining the size of the Git bare repository.

Note that I'd still allocate a certain fixed size for the cache (like we do for the asset cache) because the file system will in most cases become full for other reasons and it makes no sense to keep deleting repos from the Git cache when e.g. some big iso files actually take the space. And that's where "du" fits in. We can still use "df" to prevent any expensive cleanup logic from happening (including "du") if there's enough space on the filesystem anyway.